lib/Optimizer/Transforms/ArgumentSynthesis.cpp

```diff
@@ -147,6 +114,61 @@ class ArgumentSynthesisPass
     // substituted.
     func.eraseArguments(replacedArgs);
   }
+
+  void runOnOperation() override {
+    ModuleOp mod = getOperation();
+    MLIRContext *ctx = mod.getContext();
+
+    // 1. Collect all substitution modules.
+    std::list<std::string> funcNames;
+    DenseMap<StringRef, OwningOpRef<ModuleOp>> substModules;
+
+    for (auto &item : funcList) {
+      auto pos = item.find(':');
+      if (pos == std::string::npos)
+        continue;
+
+      std::string funcName = item.substr(0, pos);
+      std::string text = item.substr(pos + 1);
+
+      if (text.empty()) {
+        LLVM_DEBUG(llvm::dbgs() << funcName << " has no substitutions.");
+        continue;
+      }
+
+      // Create a Module with the substitutions that we'll be making.
+      LLVM_DEBUG(llvm::dbgs()
+                 << funcName << " : substitution pattern: '" << text << "'\n");
+      auto substModule = [&]() -> OwningOpRef<ModuleOp> {
+        if (text.front() == '*') {
+          // Substitutions are a raw string after the '*' character.
+          return parseSourceString<ModuleOp>(text.substr(1), ctx);
+        }
+        // Substitutions are in a text file (command-line usage).
+        return parseSourceFile<ModuleOp>(text, ctx);
+      }();
+      assert(*substModule && "module must have been created");
+
+      auto &name = funcNames.emplace_back(funcName);
+      substModules.try_emplace(name, std::move(substModule));
+    }
+
+    // 2. Merge symbols from substitution modules into the source module.
+    for (auto &[funcName, substMod] : substModules) {
```

Comment:
The substitution module should be applied to the source module atomically as a single unit.

Reply:
removed